rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,382 @@
+        return IPTag.from_bytestring(ack.data)
+
+    @ContextMixin.use_contextual_arguments
+    def iptag_clear(self, iptag, x=Required, y=Required):
+        """Clear an IPTag.
+
+        Parameters
+        ----------
+        iptag : int
+            Index of the IPTag to clear.
+        """
+        self.send_scp(x, y, 0, SCPCommands.iptag,
+                      int(consts.IPTagCommands.clear) << 16 | iptag)
+
+    @ContextMixin.use_contextual_arguments
+    def led_on(self, led, x=Required, y=Required):
+        """Turn an LED on.
+
+        Parameters
+        ----------
+        led : int
+            Number of the LED to set the state of (0-3)
+        """
+        arg1 = int(consts.LEDAction.on) << (led * 2)
+        self.send_scp(x, y, 0, SCPCommands.led, arg1=arg1, expected_args=0)
+
+    @ContextMixin.use_contextual_arguments
+    def led_off(self, led, x=Required, y=Required):
+        """Turn an LED off.
+
+        Parameters
+        ----------
+        led : int
+            Number of the LED to set the state of (0-3)
+        """
+        arg1 = int(consts.LEDAction.off) << (led * 2)
+        self.send_scp(x, y, 0, SCPCommands.led, arg1=arg1, expected_args=0)
+
+    @ContextMixin.use_contextual_arguments
+    def led_toggle(self, led, x=Required, y=Required):
+        """Toggle the state of an LED.
+
+        Parameters
+        ----------
+        led : int
+            Number of the LED to set the state of (0-3)
+        """
+        arg1 = int(consts.LEDAction.toggle) << (led * 2)
+        self.send_scp(x, y, 0, SCPCommands.led, arg1=arg1, expected_args=0)
+
+    @ContextMixin.use_contextual_arguments
+    def sdram_alloc(self, size, tag=0, x=Required, y=Required,
+                    app_id=Required):
+        """Allocate a region of SDRAM for the given application.
+
+        Parameters
+        ----------
+        size : int
+            Number of bytes to attempt to allocate in SDRAM.
+        tag : int
+            8-bit tag that can be used identify the region of memory later.  If
+            `0` then no tag is applied.
+
+        Returns
+        -------
+        int
+            Address of the start of the region.
+
+        Raises
+        ------
+        SpiNNakerMemoryError
+            If the memory cannot be allocated, or the tag is already taken or
+            invalid.
+        """
+        assert 0 <= tag < 256
+
+        # Construct arg1 (op_code << 8) | app_id
+        arg1 = (consts.AllocOperations.alloc_sdram << 8 |
+                (self.app_id if app_id is None else app_id))
+
+        # Send the packet and retrieve the address
+        rv = self.send_scp(x, y, 0, SCPCommands.alloc_free, arg1, size, tag)
+        if rv.arg1 == 0:
+            # Allocation failed
+            raise SpiNNakerMemoryError(size, x, y)
+        return rv.arg1
+
+CoreInfo = collections.namedtuple(
+    'CoreInfo', "p2p_address physical_cpu virt_cpu version "
+                "buffer_size build_date version_string")
```

Comment:
A docstring defining approximately what these values are would be appropriate here. (Note: to get namedtuple docs to appear right in sphinx, "Parameters" should be used to document them... Not ideal but still). (Also, I'm assuming you do unpack p2p_address into an (x, y) tuple).

    class CoreInfo(collections.namedtuple(
                   'CoreInfo',
                   "p2p_address physical_cpu virt_cpu version "
                   "buffer_size build_date version_string")):
        """Information returned about a core by sver.
        
        Paramters
        ---------
        p2p_address : (x, y)
            Logical location of the chip in the system.
        physical_cpu : int
            The physical ID of the core. (Not useful to most users).
        virt_cpu : int
            The virtual ID of the core. This is the number used by all high-level
            software APIs.
        version : float
            Software version number. (Major version is integral part, minor version
            is fractional part).
        buffer_size : int
            Maximum supported size (in bytes) of the data portion of an SCP packet.
        build_date : int
            The time at which the software was compiled as a unix timestamp. May be
            zero if not set.
        version_string : string
            Human readable, textual version information split in to two fields by a
            "/". In the first field is the kernal (e.g. SC&MP or SARK) and the
            second the hardware platform (e.g. SpiNNaker).
        """